src/System.Private.ServiceModel/src/System/ServiceModel/Channels/HttpResponseMessageHelper.cs

```diff
@@ -0,0 +1,367 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.ServiceModel.Diagnostics.Application;
+using System.ServiceModel.Security;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace System.ServiceModel.Channels
+{
+    internal class HttpResponseMessageHelper
+    {
+        private readonly HttpChannelFactory<IRequestChannel> _factory;
+        private readonly MessageEncoder _encoder;
+        private readonly HttpRequestMessage _httpRequestMessage;
+        private readonly HttpResponseMessage _httpResponseMessage;
+        private string _contentType;
+        private long _contentLength;
+
+        public HttpResponseMessageHelper(HttpResponseMessage httpResponseMessage, HttpChannelFactory<IRequestChannel> factory)
+        {
+            Contract.Assert(httpResponseMessage != null);
+            Contract.Assert(httpResponseMessage.RequestMessage != null);
+            _httpResponseMessage = httpResponseMessage;
+            _httpRequestMessage = httpResponseMessage.RequestMessage;
+            _factory = factory;
+            _encoder = factory.MessageEncoderFactory.Encoder;
+        }
+
+        internal async Task<Message> ParseIncomingResponse()
+        {
+            ValidateAuthentication();
+            ValidateResponseStatusCode();
+            bool hasContent = await ValidateContentTypeAsync();
+            Message message = null;
+
+            if (!hasContent)
+            {
+                if (_encoder.MessageVersion == MessageVersion.None)
+                {
+                    message = new NullMessage();
+                }
+                else
+                {
+                    return null;
```

Comment:
When the MessageVersion is None, it is acceptable to receive a response with no body as nothing dictates that you need a body. A NullMessage represents a message with no content but can still contain other metadata such as message properties. For example, HttpResponseMessageProperty containing the response status code and headers. Other message versions require a SOAP message be returned so having no body isn't acceptable and is an error condition. In those cases, representing as a NullMessage doesn't reflect the situation so null is returned which means "we didn't receive a response soap message". I don't know why null is returned instead of throwing an exception (that's what I would do if I was designing this from scratch), but this is existing behavior, the semantics of which is understood by calling code.